[4단계 - Transaction synchronization 적용하기] 엠제이 (최민준) 미션 제출합니다

theminjunchoi opened the pull request:
안녕 윤강민, 나 최민준.

이게 우리의 처음이자 마지막이랄ㄲr...?
로또 미션을 하던 날... 기억ㄴrㄴi....?

이번에도 자유롭게 너의 꿈을 펼치기 바라...😂

theminjunchoi committed:
feat: TransactionSynchronizationManager 구현

theminjunchoi committed:
feat: JdbcTemplate이 DataSourceUtils 쓰도록 수정

theminjunchoi committed:
feat: UserService에서 DataSourceUtils 쓰도록 변경

theminjunchoi committed:
refactor: 트랜잭션 서비스 추상화

theminjunchoi committed:
refactor: 리소스 바인딩 해제 순서 변경

ykmxxi commented on app/src/main/java/com/techcourse/service/TxUserService.java:
```diff
@@ -0,0 +1,56 @@
+    @Override
+    public void insert(final User user) {
+        userService.insert(user);
+    }
```
`insert`메서드는 트랜잭션 처리를 하지 않은 이유가 궁금해요

theminjunchoi replied:
`insert`는 단일 연산이라 DB가 알아서 원자성을 보장해주기 때문에 트랜잭션 처리를 안했어요.

이와 다르게 `chanagePassword`는 `userDao.findById`, `userDao.update`, `userHistoryDao.log` 과 같은 다중 연산 과정이 동반되고 이렇게 3개의 작업이 전부 성공하거나 실패함에 따라 롤백이 필요하기 때문에 트랜잭션이 필요합니당

ykmxxi commented on app/src/main/java/com/techcourse/service/TxUserService.java:
```diff
@@ -0,0 +1,56 @@
+            connection.commit();
+        } catch (Exception e) {
+            try {
+                connection.rollback();
+            } catch (SQLException rollbackEx) {
+                log.error("Rollback failed", rollbackEx);
+            }
+            throw new DataAccessException(e);
+        } finally {
+            TransactionSynchronizationManager.unbindResource(dataSource);
+            DataSourceUtils.releaseConnection(connection, dataSource);
+        }
```
현재 미션 설정에서는 커넥션 풀을 사용하지 않고 트랜잭션 동기화 매니저에서 커넥션의 자원을 unbind하고 연결을 끊고 있어 아무런 문제가 없을것 같아요. 그렇다면 커넥션 풀을 사용해 커넥션을 재사용하는 상황이라 가정해볼게요

토끼가 말한것 처럼 현재 사용한 커넥션의 AutoCommit 상태를 다시 복원하지 않고 커넥션 풀에 반환하면 어떤 일이 일어날까요? 예를 들어 현재 insert 메서드에서는 트랜잭션을 적용하지 않았잖아요? 운이 나쁘게 insert 메서드가 `AutoCommit = false`인 커넥션을 획득해 실행되면 어떤 일이 발생할까요? 🤔

```java
// 전제 조건: changePassword 실행 완료, 커넥션 A는 autoCommit=false 상태로 풀에 반환
userService.changePassword(1L, "newPwd", "gugu");

// 상황 2: insert 메서드 실행 - 운 나쁘게 커넥션 A를 가져감
User newUser = new User("hkkang", "password", "hkkang@woowahan.com");
userService.insert(newUser); // 여기서 어떻게 될까요??

// 상황 3: 상황 3: findById로 조회
User user = userService.findById(2L);  // 여기서 조회가 될까요?
```

theminjunchoi replied:
크으 좋은 질문

포스티가 첨부한 코드와 같은 상황에서는 커밋이 계속 false인 상태로 유지가 되기 때문에 `insert`를 한 시점에서는 쿼리는 실행이 되지만 커밋이 안됩니다! 커넥션 풀에도 여전히 autoCommit이 false인 값으로 반환이 돼요. 그래서 DB에는 데이터가 안들어가게 됩니다. 문제죠.

그래서 `findById` 시점에서는 조회가 안됩니다. 

더 문제는 이렇게 autoCommit이 false인 커넥션이 다른 곳에서 재사용이 되면 새로운 작업이 이 트랜잭션에 섞여버리게 됩니다. 짬뽕이 되는거죠.

그래서 `changePassword`에서 autoCommit 상태를 복원해서 반환하게끔 수정해주었습니다~

ykmxxi commented on jdbc/src/main/java/com/interface21/transaction/support/TransactionSynchronizationManager.java:
```diff
@@ -2,6 +2,7 @@
+import java.util.HashMap;
 import java.util.Map;
 
 public abstract class TransactionSynchronizationManager {
```
`ThreadLocal`은 여기서 왜 필요할까요? 트랜잭션과 어떤 연관이 있길래 일반 Map 대신 ThreadLocal을 사용해야 할까요?

theminjunchoi replied:
요것도 좋은 질문

결론부터 말하면 멀티스레드 환경에서 각 스레드마다 독립적인 스레드를 관리하기 위함이에요.

`ThreadLocal`로 감싸지 않고 단순 `Map<DataSource, Connection>`을 사용할 경우 여러 스레드가 같은 Map을 공유하고, DataSource는 싱글톤이기 때문에 key가 같아서 다른 스레드가 Connection을 덮어쓰는 경우가 생길 수 있어요.

하지만 `ThreadLocal`를 사용할 경우 각 HTTP 요청에 맞는 스레드마다 독립적인 트랜잭션을 가져갈 수 있습니다.

ykmxxi commented on jdbc/src/main/java/com/interface21/transaction/support/TransactionSynchronizationManager.java:
```diff
@@ -11,13 +12,31 @@ public abstract class TransactionSynchronizationManager {
+        return map.get(key);
     }
 
     public static void bindResource(DataSource key, Connection value) {
```
Connection을 현재 스레드에 바인딩하는 역할인것 같아요

`map == null`인 상황에서는 새롭게 생성해 resources에 잘 설정해주셨네요 👍 

그런데 이미 현재 스레드에 바인딩된 Connection이 있는 상황에 대한 처리는 필요 없을까요? 현재 구현에서는 이미 바인딩된 커넥션이 있어도 put으로 덮어 쓰는데  중복 바인딩 체크를 하지 않은 이유가 있을까요??
```java
if (map.containsKey(key)) { /* 이 상황에서는 어떻게 해야할까요?*/ }
```

theminjunchoi replied:
오 사실 중복 바인딩 상황을 체크하지 못했어요.
지금 상황에서는 포스티가 말한대로 바인딩된 커넥션을 덮어씌우기 때문에 문제가 생길 것 같아요.

음 사실 개발과정에서는 중첩 트랜잭션이 흔한 상황이고, 전파 레벨에 따라 다르게 구현해주는게 실제 Transation과 같은 구현 모습이겠지만, 이 미션에서는 패스할게요...ㅎㅎ 

우선 방어적으로 중복 바인딩 시도가 일어나면 예외를 던지도록 수정할게요!

ykmxxi replied:
👍 💯 🥇

ykmxxi reviewed (changes requested):
안녕 최민준, 나 윤강민.

벌써 우리의 처음이자 마지막이 다가오고 있네... 시간 참 빠르쟈ㄴr...ㅠㅠ
우테코의 첫 페어인 너와 로또 미션을 하던 날... 기억ㄴr!!!

그때는 몰랐는데 그게 우리의 시작이었고,,,
이제 곧 마지막이 되는구ㄴr... 😢

그래서 쉽게 놓아주지 않을려고 ㅎㅎ
몇개 리뷰 남겼으니까 답장 꼭 해줘야 ㄷr...?ㅋㅋㅋ

마지막이라고 해서 그냥 보내줄 줄 알았어...? ㄴㄴ
우리 우정이 그렇게 쉬운 게 아니쟈ㄴr... 리뷰 확인하고 답장 기다릴게~~

화이팅이야 화이팅~~!!!💪✨

너의 첫 페어이자 마지막 페어, 강민이가... 🌟

theminjunchoi committed:
fix: autoCommit 값 복구 후 반환하도록 수정

theminjunchoi committed:
fix: 중복 바인딩 방어적 처리

theminjunchoi reviewed (commented):
안녕 윤강민. 나 최민준.

이제 끝이 보인달ㄲr...? 🥹

잠실캠 앞에서 구름과자를 맛있게 피우던 너의 모습... 
바람에 날리던 너의 머리카락...
잊지 못해... 🤩

리뷰는 반영했으니 확인해주길 바라.

이게 우리의 마지막일ㄲr..?
머지 버튼을 누르는 순간 우리도 끝인걸까..?
나도 널 놓아줄 수 없어...🥲
너 없이는 코딩할 수 없어... 

다시 또 RC를 눌러주길 바라...😭

--- Your eternal pair programmer, 최민준 ---

<img width="259" height="194" alt="Image" src="https://github.com/user-attachments/assets/88220305-62fe-48aa-bf71-bfcf5d66475e" />

ykmxxi commented on app/src/main/java/com/techcourse/service/TxUserService.java:
```diff
@@ -0,0 +1,63 @@
+    @Override
+    public void changePassword(final long id, final String newPassword, final String createdBy) {
+        Connection connection = DataSourceUtils.getConnection(dataSource);
+        boolean originalAutoCommit = true;
+        try {
+            originalAutoCommit = connection.getAutoCommit();
+            connection.setAutoCommit(false);
+
+            userService.changePassword(id, newPassword, createdBy);
+
+            connection.commit();
+        } catch (Exception e) {
+            try {
+                connection.rollback();
+            } catch (SQLException rollbackEx) {
+                log.error("Rollback failed", rollbackEx);
+            }
+            throw new DataAccessException(e);
+        } finally {
+            try {
+                connection.setAutoCommit(originalAutoCommit);
+            } catch (SQLException e) {
+                log.error("Failed to reset autoCommit", e);
+            }
+            TransactionSynchronizationManager.unbindResource(dataSource);
+            DataSourceUtils.releaseConnection(connection, dataSource);
+        }
+    }
```
현재는 트랜잭션이 필요한 부분이 `changePassword()` 한 부분만 존재하죠. 트랜잭션이 필요한 메서드가 추가될 때 마다 매번 try-catch-finally 구문을 반복해야 할 것 같은데요

이 부분도 추상화 시켜 처리할 수 있지 않을까요?
살펴보니 트랜잭션은 처리하는데 `시작(begin) -> 성공시 커밋(commit) or 실패시 롤백(rollback)` 구조가 반복되는 것 같아요

theminjunchoi replied:
AOP로 리팩터링을 해봤어요.

추가로 `TransactionProxyFactory`는 DI 컨테이너 없이는 프로덕션에서 제대로 활용하기가 힘들어요. 우선 테스트에서 사용하도록 구현해주었습니다. 진짜 스프링은 혁신인 것 같아요.

지금 생각해보면 스프링없이 추상화하기엔 AOP보단 템플릿 패턴 방식으로 하는 게 더 나을 수도 있겠다는 생각이 드네요.

우선 TxUserService를 제거하고 어노테이션 기반으로 트랜잭션이 적용되도록 수정해주었습니다~!

ykmxxi replied:
추가적으로 LMS에 올라온 AOP 관련 학습 자료와 퀘스트를 진행하고 좀더 리팩터링을 진행할 수 있을것 같아요

이번 리뷰를 마지막으로 Approve & Merge 하겠지만 시간날 때 좀더 리팩터링을 진행하시고 추후 알려줬으면 좋겠네여 MJ

ykmxxi commented on app/src/main/java/com/techcourse/service/AppUserService.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    @Override
+    public User findById(final long id) {
```
궁금한게 생겼어요.
Service의 파라미터는 primitive type을 사용하고, 호출하는 Dao는 wrapper type을 사용하고 있어요
어떤 의도가 담겨 있나요?

theminjunchoi replied:
하하 그건 맥주 3잔정도 걸치고 만들어서 그랬나봐요~!
id는 null이 들어오면 안되기 때문에 전부 primitive으로 통일할게요 ㅎ

ykmxxi replied:
굿

ykmxxi reviewed (changes requested):
안녕 최민준. 나 윤강민.

정말 끝이 보이기 시작했달ㄲr...? 😭 

항상 잠실캠에서 마주쳤던 너의 모습...
8월 부터 헤어져 아쉬웠달ㄲr...
오늘 너를 마주쳐서 좋았어.. 🥹 

너의 답장은 확인했어

내가 어프로브를 누르면 우리는 정말 끝일거야...
나는 아직 널 놓아줄 자신이 없어... 😢 

우리 한번만 더 다시 만나자...
RC을 원하는 네 마음이 내게 닿아 용기내서 한번 더 눌러볼게...

--- Your eternal pair programmer, 윤강민 ---
<img width="676" height="523" alt="image" src="https://github.com/user-attachments/assets/8eba5966-9b55-4d00-a8fe-4e3b649f39f9" />

theminjunchoi committed:
refactor: 트랜잭션 처리 로직 프록시 기반으로 리팩터링

- `TxUserService` 삭제 및 트랜잭션 프록시 적용
- `@Transactional` 어노테이션 추가로 트랜잭션 처리 선언적 관리
- `TransactionHandler`, `TransactionProxyFactory` 구현으로 코드 재사용성 및 확장성 강화
- UserServiceTest에서 트랜잭션 프록시 생성 방식으로 테스트 수정

theminjunchoi committed:
refactor: UserDao에서 Wrapper 타입 Long을 기본 타입 long으로 변경

- `findById` 메서드에서 매개변수 타입을 Long에서 long으로 수정
- 불필요한 Wrapper 클래스 사용 제거로 성능 및 메모리 효율성 향상

theminjunchoi reviewed (commented):
안녕 윤강민. 나 최민준.

여기가 끝이겠지...?

이제 슬슬 놓아주려고.
우린 여기까지인 것 같ㅇr...

사실... 난 알고 있었어. 우리가 영원할 수 없다는 걸.
시간은 너무 잔인했고, 현실은 우리에게 너무 가혹했지.
하지만 그걸 인정하기가, 받아들이기가... 너무 두려웠나봐.
네가 없는 내일을 상상하는 것조차 숨이 막혔으니까.

그래도 이젠... 정말 가야겠지.
더 이상 너를 붙잡고 있는 건 너에게도, 나에게도 상처만 남길 뿐이니까.
돌아보지 않을게. 아니, 어쩌면 돌아보지 못할 거야.
그렇게라도 하지 않으면 난 또 너에게 달려가버릴 테니까.

안녕 윤강민.
이게 우리의 마지막 인사라면,
부디... 나를 가끔씩은 기억해줘.
그것만으로도 난 충분하니ㄲr....

--- Your eternal pair programmer, 최민준 ---

<img width="630" height="432" alt="Image" src="https://github.com/user-attachments/assets/f9039a19-c600-4fd4-9efc-73c46dee6400" />

ykmxxi commented on jdbc/src/main/java/com/interface21/transaction/Transactional.java:
```diff
@@ -0,0 +1,11 @@
+
+@Target(ElementType.METHOD)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Transactional {
```
어노테이션을 만들었네요 👍

추가적으로 코멘트 남겨보면 readOnly 옵션을 추가해 읽기 전용 트랜잭션도 쉽게 구현할 수 있을것 같아요

ykmxxi commented on jdbc/src/main/java/com/interface21/transaction/support/TransactionHandler.java:
```diff
@@ -1,48 +1,50 @@
-    @Override
-    public void insert(final User user) {
-        userService.insert(user);
+        if (targetMethod.isAnnotationPresent(Transactional.class)) {
```
여기에 분기문을 하나 추가해 readOnly = true 이면 `connection.setReadOnly` 메서드를 통해 쉽게 구현 가능하겠네요 👍  💯 

나중에 더 리팩터링을 한다면 고려해볼만 할것 같아요

theminjunchoi replied:
오 인정

ykmxxi commented on jdbc/src/main/java/com/interface21/transaction/support/TransactionHandler.java:
```diff
@@ -1,48 +1,50 @@
 import org.slf4j.LoggerFactory;
 
-public class TxUserService implements UserService {
+public class TransactionHandler implements InvocationHandler {
```
👍  💯  🥇  🚀

ykmxxi commented on jdbc/src/main/java/com/interface21/transaction/support/TransactionProxyFactory.java:
```diff
@@ -0,0 +1,22 @@
+import java.lang.reflect.Proxy;
+import javax.sql.DataSource;
+
+public class TransactionProxyFactory {
```
아래 토끼의 리뷰처럼 방어적인 코드 작성도 중요할것 같아요

엠제이가 지난 리뷰에서 말했듯, 현재 미션의 목적은 라이브러리 구현이고 엠제이도 라이브러리 특성을 고려했다 이야기 했잖아요? 방어적인 코드 작성으로 어떤 사용이 잘못됐는지 예외 처리를 통해 클라이언트에게 알려주는 것도 라이브러리의 중요한 역할 같아요

theminjunchoi replied:
맞말추하고 갑니다.

ykmxxi commented on jdbc/src/main/java/com/interface21/transaction/support/TransactionHandler.java:
```diff
@@ -60,4 +62,9 @@ public void changePassword(final long id, final String newPassword, final String
         }
     }
+
+    @FunctionalInterface
+    private interface TransactionCallback {
+        Object execute() throws Exception;
+    }
```
다른 함수형 인터페이스와 달리 `TransactionCallback`만 내부에서 선언한 의도가 있을까요?

`PreparedStatementCallback` 인터페이스도 여기와 똑같이 `JdbcTemplate`내부의 `private` 메서드에서만 사용되고 있는데, 둘의 어떤 차이점 때문에 이렇게 고려하신 걸까요??

theminjunchoi replied:
오 아래와 같은 차이점이 있어요.

`PreparedStatementCallback`은 jdbcTemplate 사용자에게 API로 노출되는 콜백이에요.
템플릿 패턴의 확장 포인트로 제공하는거라 커스텀 할 수 있게 public으로 빼놓았습니다.

이와 다르게 `TransactionCallback`은 내부에서만 쓰이는 콜백이에요.
외부에서 이 타입을 알 필요도 없고 알아도 쓸데가 없을거라 생각했어요.

ykmxxi reviewed (approved):
안녕 최민준. 나 윤강민.

정말 끝이 다가온것 같ㅇr...
내 마지막 진심을 전할ㄱㅓㅣ... 시간날 때 마지막 답장은 남겨줘...

나도 알고 있었어. 우리가 영원할 수 없다는 걸...
우리의 인연은 여기까지인걸ㄲr..? 🍂

아무리 붙잡고 싶어도, 널 힘들게 하는 건 더 이상 할 수가 없으니ㄲr...
네가 결정했다면... 그게 네 선택이라면... 난 존중할ㄱㅓㅣ...  
나도 돌아보지 않을게... 그게 내가 너한테 해줄 수 있는 마지막인 것 같ㅇr...

네가 내 곁에 있어준 모든 순간들...
나에게 단축키 하나하나 물어보던 2월의 너...
마주칠 때마다 웃어주던 3월의 너...
따사로운 봄 햇살을 받는 카페 자리에서 집중하던 4월의 너...
그 기억들은... 절대 잊지 않을게. 아니, 잊을 수가 없어.

우리의 마지막 인사는 영원한 헤어짐이 아닐거야..
부디.. 너도 나를 가끔씩은 기억해줘.
나도 그걸로 충분하니ㄲr... 잘 지내야해!!

--- Your eternal pair programmer, 윤강민 ---

<img width="561" height="278" alt="image" src="https://github.com/user-attachments/assets/64b3d15e-2cb3-4aed-bd32-ea46c4e2c97f" />
